qcat: use direct memcpy when all the inputs and output share the same scale and zero_point

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74999
* #72575
* __->__ #71903

Differential Revision: [D34113897](https://our.internmc.facebook.com/intern/diff/D34113897)

mingfeima committed:
qcat: use direct memcpy when all the inputs and output share the same scale and zero_point

[ghstack-poisoned]

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"




[ghstack-poisoned]

vadimkantorov commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -41,6 +41,18 @@ void check_tensor_memory_format(const Tensor& ref, const Tensor& other) {
+    int64_t zero_point) {
+  bool is_valid = true;
+  for (const auto i : c10::irange(qxs.size())) {
+    is_valid |= qxs[i].q_scale() == scale;
```
if `is_valid` is initialized as true, then any further `|=` will not change its value. Did you mean `&=`?

mingfeima replied:
Ohhh, yes! Fixed.

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"




[ghstack-poisoned]

mingfeima commented:
TODO: this PR won't cover the scenario from shufflenet since:
* memory format is not channels last since use of `.contiguous()` from channel_shuffle func

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"




[ghstack-poisoned]

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"




[ghstack-poisoned]

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"




[ghstack-poisoned]

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"




[ghstack-poisoned]

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"




[ghstack-poisoned]

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"




[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34113897).

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"


Differential Revision: [D34113897](https://our.internmc.facebook.com/intern/diff/D34113897)

[ghstack-poisoned]

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"


Differential Revision: [D34113897](https://our.internmc.facebook.com/intern/diff/D34113897)

[ghstack-poisoned]

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"


Differential Revision: [D34113897](https://our.internmc.facebook.com/intern/diff/D34113897)

[ghstack-poisoned]

mingfeima commented:
### Updates of this PR
originally, qcat would go dequant-cat-quant. This PR checks whether the input has the same qparam with output, if yes use a memcpy instead of dequant/quant.

for performance update, please check the previous PR.

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"


Differential Revision: [D34113897](https://our.internmc.facebook.com/intern/diff/D34113897)

[ghstack-poisoned]

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"


Differential Revision: [D34113897](https://our.internmc.facebook.com/intern/diff/D34113897)

[ghstack-poisoned]

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"


Differential Revision: [D34113897](https://our.internmc.facebook.com/intern/diff/D34113897)

[ghstack-poisoned]

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"


Differential Revision: [D34113897](https://our.internmc.facebook.com/intern/diff/D34113897)

[ghstack-poisoned]

jerryzh168 commented:
is this connected with quantization flow?

mingfeima commented:
> is this connected with quantization flow?

it's not related to quantization flow, only a fast path on the quantized kernel.

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"


Differential Revision: [D34113897](https://our.internmc.facebook.com/intern/diff/D34113897)

[ghstack-poisoned]

VitalyFedyunin commented:
@frank-wei  can you please commandeer this PR internally and take care of landing.

mingfeima committed:
Update on "qcat: use direct memcpy when all the inputs and output share the same scale and zero_point"


Differential Revision: [D34113897](https://our.internmc.facebook.com/intern/diff/D34113897)

[ghstack-poisoned]

frank-wei commented:
@frank-wei has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34113897).
